### ISSUE TITLE
Hierarchical albums

### ISSUE THREAD

**Issue by bobobo1618** [photoprism/photoprism#401]
I think a way to navigate a hierarchy of collections is essential. A couple of use-cases:

- Breaking down large collections into smaller ones, e.g. 2020 -> Europe Holiday -> Paris
- Allowing intuitive navigation of existing on-disk collections of photos (which are frequently organised into a folder hierarchy)
- 1:1 mirrors of other software that has an album hierarchy (Lightroom, Piwigo etc.)

I have some thoughts on how to implement it but is this something that fits with what PhotoPrism is intended to do?

**Comment by lastzero** [comment]
You can already browse by folder: https://demo.photoprism.org/library/files

Did you see, that we also have categories for organizing albums?

I'm not a big fan of hierarchies to be honest, I'm more like an SQL guy who thinks in sets and filters. That was a big, very successful innovation back then. But I understand some people want their trees back :)

Album selection and permissions also get overly complex with nested albums. When there are different namespaces (parent albums), a name can be used many times. Building a selection control without ambiguities for this, that works well on mobile, can be a hell of a nightmare. I've developed such a solution for one of the world's largest telecommunications companies in the past, see https://github.com/photoprism/photoprism/wiki/Mediencenter.

![](https://dl.photoprism.org/assets/screenshots/mediencenter/filemanager.jpg)

**Comment by bobobo1618** [comment]
Hmm, browsing by folders doesn't appear to work for me for some reason, PhotoPrism just tells me "no results". I'll figure out how to get PhotoPrism to log the SQL query and figure out what's going on.

SQL and trees aren't mutually exclusive, CTEs are fun!

That's a good point, maybe nested albums aren't such a great idea. Folders are the most important thing and it looks like when things are working as intended, you already have those.

**Comment by lastzero** [comment]
Maybe we need to tweak / fix the file browser query for Sqlite, thought it is working. If not, related tests should be reworked too. You're welcome to create an issue for this.

**Comment by bobobo1618** [comment]
Not SQLite this time but symlinks. #403 :)

**Comment by lastzero** [comment]
Our demo uses sqlite too, so it should work :)

Symlinks are generally supported, but maybe not listed / indexed as folder. Probably not difficult to add.

**Comment by JohanAR** [comment]
I think you could achieve a similar result by adding a "group by" option for albums, so instead of immediately showing individual files it would let you narrow down the selection.

It could for example let you choose the following grouping: album, country, year, year+month, date

If you configure your "Europe Holiday" to group by album it would show "Paris" as if it was a sub-folder, as long as you have added photos to both albums. The only difference from an actual hierarchy is that "Paris > Europe Holiday" would be equivalent to "Europe Holiday > Paris", but I think it fits better with the database view of your collection.

**Comment by wiwie** [comment]
I would also like this feature. More precisely, I would like to be able to share a large collection of photos with a grouping structure / partitioning reflected. I don't know about others, but this is a very common use-case I have.

In this context, commenting on the suggestion to use album categories: Categories do not seem to support sharing, so if I want to share a large collection of photos that are naturally partitioned somehow and I want this to be reflected in the share, I can't use categories, correct?

Generally, I haven't found any way to do the use-case I described in photoprism.

**Comment by titoko** [comment]
I've got the same use case:
I want to give two share link, one for Family one for friend leading to multiple album (which could changed) so i don't have to share a new link each time I create new album

**Comment by prov3it** [comment]
I'd like to see this feature as well. I've got thousands of scanned negatives which need subalbums to know in what physical folder i can find them.

**Comment by graciousgrey** [comment]
@prov3it this might solve your problem: https://github.com/photoprism/photoprism/issues/1536

**Comment by P-Verbrugge** [comment]
Same here. I am giving Photoprism a testrun to see if can use it as a replacement for PhotoStation on my NAS. I have ordered all my pictures by location and year subfolders. For example, a vacation folder has 2 subfolders: 2020 and 2021. The pictures are stored in those folders. I would like to make albums like that too. Now it's all dropped in one album.

Something similar like Photoshow does. Photoshow is a lot less advanced, but it uses the folder structure as album structure and keeps your own structure intact.

**Comment by lastzero** [comment]
You can create albums from file system folders using the context menu. The hierarchy is visible in Library > Originals. This issue is about also showing it in the Folders view in addition to searching by name. Photos from separate folders should not be "dropped" all in one folder. Can you give us an example?

**Comment by P-Verbrugge** [comment]
@lastzero 

"The hierarchy is visible in Library > Originals." 

I noticed that. But I want it also to be visible with the "albums".

What I mean with "dropped in one folder" is. When I have a folder "Vacation". In that folder I have 2 subfolders. 2020 and 2021. I can see that structure as you said under "Library -> Originals". That's good. 
When I add "Vacation" with the context menu to albums, I notice that the folder structure is not visible in the album and all photo's (that are located in subfolders) are dropped in one folder (album) called "Vacation".

I guess the idea is not to show the same storage structure for albums as for orginals. Though, it would still be a nice feature to be able to add sub-albums to albums.

**Comment by lastzero** [comment]
> I noticed that. But I want it also to be visible with the "albums".

Folders are visible in Folders, not Albums (so you don't need to manually create an Album for every Folder):

https://demo.photoprism.org/folders

Albums are not nested and manually managed. If you add files to a Folder, you probably expect the Folder contents to be updated automatically, not manually?

> What I mean with "dropped in one folder" is. When I have a folder "Vacation". In that folder I have 2 subfolders. 2020 and 2021. I can see that structure as you said under "Library -> Originals". That's good.

> When I add "Vacation" with the context menu to albums, I notice that the folder structure is not visible in the album and all photo's (that are located in subfolders) are dropped in one folder (album) called "Vacation".

Because the selection is recursive. Creating an album from a folder will always create exactly one album with all files in the folder including all subfolders. It's like Windows Explorer: You delete, copy, or move a folder and all the files in its subfolders will be affected as well.

> I guess the idea is not to show the same storage structure for albums as for orginals. Though, it would still be a nice feature to be able to add sub-albums to albums.

A major issue here is mobile UX as deeply nested, complex structures are not very user friendly on mobile devices. I think it makes sense to support / show nested folders when they are nested in the filesystem but don't use trees all over the place. This will also greatly simplify multi-user support as you don't need to inherit permissions through different levels. ACLs in Microsoft Windows are a complete mess IMHO:

https://espace.cern.ch/winservices-help/NICESecurityAndAntivirus/NICESecurityHowTo/Pages/ManagingACLSettingPermssion.aspx

There are other ways to find what you are looking for, like search filters, labels, keywords, or tags. Those also don't require you to move files around in the filesystem if you want to sort / arrange them differently.

**Comment by P-Verbrugge** [comment]
Thanks for explaining. It makes the design choices more clear. I will look into tags, etc then.

**Comment by joachimtingvold** [comment]
@wiwie wrote:
> I would also like this feature. More precisely, I would like to be able to share a large collection of photos with a grouping structure / partitioning reflected (let me know whether that should be a separate issue). I don't know about others, but this is a very common use-case I have.

@titoko wrote:
> I've got the same use case: I want to give two share link, one for Family one for friend leading to multiple album (which could changed) so i don't have to share a new link each time I create new album.

That usecase would technically be covered via #1162 (so no need for hierarchical albums).

**Comment by wiwie** [comment]
> @wiwie wrote:
> 
> > I would also like this feature. More precisely, I would like to be able to share a large collection of photos with a grouping structure / partitioning reflected (let me know whether that should be a separate issue). I don't know about others, but this is a very common use-case I have.
> 
> @titoko wrote:
> 
> > I've got the same use case: I want to give two share link, one for Family one for friend leading to multiple album (which could changed) so i don't have to share a new link each time I create new album.
> 
> That usecase would technically be covered via #1162 (so no need for hierarchical albums).

not really. it may depend on how flexibly that feature is implemented, but it sounds like it would be bound to depths of 1. I'd like to be able to share a hierarchy of depths>1. fx year/month.

of course hierarchies could always be flattened to depth=1. but for hierarchies with many childs that would ruin the usability IMO, e.g. to stick with my example finding a specific month of a specific year may be cumbersome.

**Comment by cgrauer** [comment]
Even if time has passed... 

Folders are good for hierarchical guys, but refer only 1:n and they are not manageable from within PhotoPrism. 
Albums are editable, refer n:n, but they are flat (also Categories only add 1 extra level). If I would use Albums for grouping my fotos, I would end up with about 439 Albums... 
Labels (which I'd normally prefer...) refer n:n, but you have to add fotos one by one to labels and you can't filter by multiple labels (what makes labels useful in the first place). And labels are being created out of nowhere...
Categories - just falling from the sky.
Keywords same as Labels, only no gallery view. 
In my opinion, many different attempts but none are consistent.

If we could add fotos to labels like we can to albums, and if we could filter by multiple labels, everyone could be happy, the SQL sets and filter guy as well as the hierarchy guy. Or at least filtering by multiple Albums and we're able to find our photos, even if it's not the common purpose of "Albums".

**Comment by graciousgrey** [comment]
> If we could add fotos to labels like we can to albums, and if we could filter by multiple labels, everyone could be happy, the SQL sets and filter guy as well as the hierarchy guy. Or at least filtering by multiple Albums and we're able to find our photos, even if it's not the common purpose of "Albums

Thank you for sharing your use case with us!

Batch editing is planned and will allow you to add multiple photos at once to a label once implemented. Also note that all labels are automatically added as keywords and the search filters for keywords and albums support AND and OR.

https://docs.photoprism.app/user-guide/search/filters/

**Comment by cgrauer** [comment]
> Also note that all labels are automatically added as keywords and the search filters for keywords and albums support AND and OR.

Right, thanks for that hint! However, I would classify this as more of a workaround, especially on mobile phones. It is not the type of handling that primarily determines the conception of a GUI. I'm not complaining, just want to share user perspective and needs for future development...

**Comment by TurboStreetCar** [comment]
Id really like to see albums within albums be possible. 

Its not super intuitive to use the folder structure, and if using the folder structure is "the way" to do this, then making albums doesn't seem to be relevant at all.

I have multiple projects, and I like to document my work with pictures. 

For example, if I do some work on my house, maybe i do a bathroom renovation, then i replace windows, then i put down new flooring.....

Its very clutter-ing to have individual albums for 10-20 different projects all on the main albums page. It would be much easier and more organized to place them within a "Home Projects" album. Being as all of these projects are related to "Home Improvements" I find myself with "Home Improvement Bathroom", "Home Improvement Kitchen", "Home Improvement Basement"...ect...ect...

Same goes for vacations/trips. 

Instead of having 10-20 albums to scroll through, it would be super convenient to place all vacation albums into a "vacation" album. This way when I'm looking for pictures from a car show, I don't have to sort through 100 recursive albums of similar but different subjects.

When you have 7k+ images and videos, it can become a real pain to scroll through everything. This is one really big hang-up I have on an otherwise pretty cool software, that may just have me stick with windows folders on my computer.

**Comment by dcarbone** [comment]
Some thoughts and a use case:

1. An album is a pointer to other things. 
2. Things referenced in one album may be referenced in other, sibling albums already. 
3. Albums do not own the things they point to.
4. Technically, there is no reason an album cannot exist as a child of another album, regardless of the concrete things they each point to.
5. Albums are NOT fixed directories!

I do NOT want albums to mimic directories as the entire point of an album is to logically group images regardless of their physical position on disk. 

I organize ALL photos by date, which makes it easier to narrow down to a particular picture / set of pictures by overall date when viewing _all_ pictures, but one of the primary purposes of an album is to point to a specific subset of those pictures with some kind of common thread to them.  The nature of this thread is arbitrary and may or may not have anything to do with date or physical location of image on disk, but may very well be relative to both. 

A use case is as follows:

I have an album with wedding pictures from multiple sources spanning multiple, sub-events of the overall wedding event. I'd like to create logical sub-groupings of the pictures from each sub-event into its own album under the overall parent album of "Wedding". This will provide an easier to use overall experience for people I share the album with.

**Comment by WhatIsThisAllAboutSilly** [comment]
Like to echo this, looking at options that aren't Piwigo/Coppermine and came across Photoprism.

The lack of nested albums and the ability to make users/groups seems like a fundamental requirement for gallery apps.

I'm currently self-hosting photos from family events, friend events and our holidays and obviously who can see what album group is one of the requirements.

Is this happening (I've subbed for the moment, but I hadn't realised that something as basic as nested albums and user management wasn't there yet) or do I continue my search/stick with Piwigo?

**Comment by davejhahn** [comment]
I understand there are potential challenges to implementing, but ultimately if part of the mission is "provide the most user- and privacy-friendly solution to keep your pictures organized and accessible.", not having the flexibility to logically group photos differently than the physical structure in a way that makes sense to the user misses that mark.  I think the same can be said about inability to modify properties in bulk (e.g. tags/keywords, location).   Additionally, I think having the ability to write back the properties to the file is crucial for being able to use images in a future platform is crucial.  

I have multiple sets of photos, spanning back over 200 years, and I really cannot find a usable way to pull them together and present in a way that doesn't feel like 10,000's of photos at a flat level or require someone to sift through 1000's of albums.  Even excluding prior generations of photos, my personal set is 45K and presents the same problem.

I have a folder structure that is Year->Month->Topic.   If someone wanted to use folders to look for "that trip to Niagara Falls", they would have to know the year to use the folder structure.  Trust me, as years go by, nobody remembers what year or month these things happened in.  

Other current options I see are:
- Searching.  Assuming that "Niagara Falls" is the folder name or tagged as such, it would present all the images.  However, I have a trip from one I was a kid to there and then a recent one with my family, so it would show all of them, and as interesting as it might be to someone in my family to see my childhood vacation, it's likely not what they would be interested in.  Also, searching never feels absolute to me versus a defined logical structure.  IS everything presented that should be and does it include things that should NOT be?
- Creating albums with some sort of prefix in place of what the hierarchy would be, e.g. "Vacations-Niagara Falls-1975", "Vacations-Niagara Falls-2022" but that would be a challenge to navigate and I'm thinking would be 100's of albums possibly 1000's when looking across my entire set, which would be quite messy.

Example of how I would use nested albums:

Vacations
  -> Niagara Falls
          -> 1975
          -> 2020
  -> California
         -> San Francisco
         -> San Jose
               -> 2000
               -> 2010
  -> Ohio

And so on.  "Vacations" is just one example, there are endless.  But above shows the flexibility--e.g. not "needing" to have a city or a year if it doesn't make sense because there was only one trip, or if it spanned multiple cities but isn't super relevant.

I mentioned the writing properties/metadata back to the files, just wanted to elaborate on that a bit.  Regardless of how great the tool is or how much better it becomes, there will be, without a doubt, a day that comes where it ceases to exist.  Might be 2 years, 5 years, 15 years, who knows.  But when you are talking about photos, it's not just some guy with his computer, but it is a historical reference, set of work, whatever.  Even when a platform ceases to exist, the images will live on, presumably forever.  Not being able to capture all the fantastic information the platform offers at the file level means that without the platform, they are exactly the same as before being introduced to the platform without any of the extended details it offered.   If there some incredible  future platform is developed, presumably with a focus on AI, having all of those details would be an incredible starting point versus starting over.  I would add that starting over will never be as good--future generations will not have the knowledge to be able to organize photos from before they were even born in any meaningful way.

I love what you guys have done, and if I sound like I am just being negative, that wasn't my intent.  I have used pretty much EVERY tool that is out there for photo management and with each one there is something great and something missing that is significant enough to make it not really an option.  Just trying to throw some ideas.

Awesome work though!

**Comment by WhatIsThisAllAboutSilly** [comment]
Almost a year has passed since I tried Photoprism - is this feature coming or not?

Some sort of clarity - either way, would be useful.

**Comment by lastzero** [comment]
We appreciate your feedback! Our team is currently finishing work on the frontend upgrade and lightbox enhancements:

- https://github.com/photoprism/photoprism/issues/3168
- https://github.com/photoprism/photoprism/issues/1307
- https://github.com/photoprism/photoprism/issues/4698
- https://github.com/photoprism/photoprism/issues/4704
- https://github.com/photoprism/photoprism/issues/4058
- https://github.com/photoprism/photoprism/issues/4762
- https://github.com/photoprism/photoprism/issues/4763

Any help in testing the preview build is very welcome and will help us to release it faster:

- https://docs.photoprism.app/release-notes/
- https://docs.photoprism.app/getting-started/updates/#development-preview

Upcoming releases will focus on multi-library support, batch editing, and AI improvements:

- https://github.com/photoprism/photoprism/issues/271
- https://github.com/photoprism/photoprism/issues/98
- https://github.com/photoprism/photoprism/issues/3124
- https://github.com/photoprism/photoprism/labels/ai

While we also plan to add support for hierarchical tags and albums, the workload of adding these features does not allow us to work on all of this at the same time.

Once we have released the features that are already in development, we may be able to give you a more specific answer as to when it will be available. Also, please note that [our software is freely available](https://www.photoprism.app/oss/faq) and most users do not pay for our work - so we need either your patience or your contribution. Thank you very much! ❤

**Comment by dannyadair** [comment]
I think maybe an easier compromise for this could be hierarchical _categories_ instead.
A photo in an album of a subcategory would automatically also be in the parent categories and found there.
